third_party/terraform/utils/retry_transport.go

```diff
@@ -0,0 +1,195 @@
+
+// NewTransportWithDefaultRetries constructs a default retryTransport that will retry common temporary errors
+func NewTransportWithDefaultRetries(t http.RoundTripper) *retryTransport {
+	return &retryTransport{
+		retryPredicates: defaultErrorRetryPredicates,
+		internal:        t,
+	}
+}
+
+// Helper method to create a shallow copy of an HTTP client with a shallow-copied retryTransport
+// s.t. the base HTTP transport is the same (i.e. client connection pools are shared, retryPredicates are different)
+func ClientWithAdditionalRetries(baseClient *http.Client, baseRetryTransport *retryTransport, predicates ...RetryErrorPredicateFunc) *http.Client {
+	copied := *baseClient
+	if baseRetryTransport == nil {
+		baseRetryTransport = NewTransportWithDefaultRetries(baseClient.Transport)
+	}
+	copied.Transport = baseRetryTransport.WithAddedPredicates(predicates...)
+	return &copied
+}
+
+// Returns a shallow copy of the retry transport with additional retry
+// predicates but same wrapped http.RoundTripper
+func (t *retryTransport) WithAddedPredicates(predicates ...RetryErrorPredicateFunc) *retryTransport {
+	copyT := *t
+	copyT.retryPredicates = append(t.retryPredicates, predicates...)
+	return &copyT
+}
+
+type retryTransport struct {
+	retryPredicates []RetryErrorPredicateFunc
+	internal        http.RoundTripper
+}
+
+// RoundTrip implements the RoundTripper interface method.
+// It retries the given HTTP request based on the retry predicates
+// registered under the retryTransport.
+func (t *retryTransport) RoundTrip(req *http.Request) (resp *http.Response, respErr error) {
+	// Set timeout to default value.
+	ctx := req.Context()
+	var ccancel context.CancelFunc
+	if _, ok := ctx.Deadline(); !ok {
+		ctx, ccancel = context.WithTimeout(ctx, defaultRetryTransportTimeoutSec*time.Second)
+	}
+
+	attempts := 0
+	backoff := time.Millisecond * 500
+
+	log.Printf("[DEBUG] Retry Transport: starting RoundTrip retry loop")
+Retry:
+	for {
+		log.Printf("[DEBUG] Retry Transport: request attempt %d", attempts)
+
+		// Copy the request - we dont want to use the original request as
+		// RoundTrip contract says request body can/will be consumed
+		newRequest, copyErr := copyHttpRequest(req)
+		if copyErr != nil {
+			respErr = errwrap.Wrapf("unable to copy invalid http.Request for retry: {{err}}", copyErr)
+			break Retry
+		}
+
+		// Do the wrapped Roundtrip. This is one request in the retry loop.
+		resp, respErr = t.internal.RoundTrip(newRequest)
+		attempts++
+
+		retryErr := t.checkForRetryableError(resp, respErr)
+		if retryErr == nil {
+			log.Printf("[WARN] Retry Transport: Stopping retries, last request was successful")
+			break Retry
+		}
+		if !retryErr.Retryable {
+			log.Printf("[WARN] Retry Transport: Stopping retries, last request failed with non-retryable error: %s", retryErr.Err)
+			break Retry
+		}
+
+		log.Printf("[DEBUG] Retry Transport: Continue retry after waiting for  %s", backoff)
+		select {
+		case <-ctx.Done():
```

Comment:
I don't think this will be a problem, but it's worth calling out. I'm not positive without some testing, but I'm pretty sure this is going to catch the context timeout but also any context cancellation that happens up stream as well. Since we are [plumbing through](https://github.com/terraform-providers/terraform-provider-google/blob/master/google/provider.go#L892) the context that allows for Terraform binary cancellation to be propagated this will catch that StopContext call. Which will mean that container cluster polling (if container cluster opts into this behavior) will need to change it's cancellation behavior [as it does right now](https://github.com/terraform-providers/terraform-provider-google/blob/master/google/container_operation.go#L68-L79). Since the resource itself is [checking for a cancellation](https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_container_cluster.go#L924-L927) I think we can just pull out the operation logic and also since a retriable timeout might also be worth "resuming" we may not have to adjust the resource logic. And also the "resume" should only happen a single time as the operation_id should get [cleared after the first read call](https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_container_cluster.go#L1001) it shouldn't cause infinite resumes.

All of this is to say: I think this is fine but when we move this logic globally or to container cluster, there will be dragons. Maybe worth adding a TODO near wherever container cluster will end up opting in?

Reply:
Synced offline - the TL;DR is that as long as we're careful with removing retryTimeDuration and we don't use a context that isn't derived from config.Context, this shouldn't cause an issue.